### PR TITLE
add Status data structure

### PR DESCRIPTION
**Pull request by zhjwpku**
This is adapted from Apache Arrow, which a pre-step before integrating the Result data structure.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -56,6 +64,11 @@ add_subdirectory(api)
 add_subdirectory(src)
 
 if(ICEBERG_BUILD_TESTS)
+  fetchcontent_declare(googletest
```
Comment (pitrou) [inline CMakeLists.txt]: This adds a GoogleTest dependency. Was there a decision to use GT? A popular alternative is https://github.com/catchorg/Catch2/, I'm not sure which one would be better.

File: api/iceberg/util/string_builder.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+#include "visibility.h"
+
+namespace iceberg {
+
+namespace util {
+
+namespace detail {
+
+class ICEBERG_EXPORT StringStreamWrapper {
```
Comment (pitrou) [inline api/iceberg/util/string_builder.h]: Do we really to keep this from Arrow (which was C++11 compatible back then), or do we want to adopt a modern formatting library such as https://en.cppreference.com/w/cpp/utility/format or a backport thereof?

**Comment by pitrou** [comment]
Also @zhjwpku can you make sure you proof-read your PR description? There are spell checkers that can help.

File: CMakeLists.txt (hunk 1 above)
Comment (zhjwpku) [reply CMakeLists.txt]: Yeah, I see that Catch2 has TDD and BDD, which is better, I will do some work and find out which one fits us better.

File: api/iceberg/util/string_builder.h (hunk 2 above)
Comment (wgtmac) [reply api/iceberg/util/string_builder.h]: I'm not sure whether `std::format` has been fully implemented by popular compilers. At least we can choose `fmt::format`: https://fmt.dev/

File: api/iceberg/status.h [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+//
+// A Status encapsulates the result of an operation.  It may indicate success,
+// or it may indicate an error with an associated error message.
+//
+// Multiple threads can invoke const methods on a Status without
+// external synchronization, but if any of the threads may call a
+// non-const method, all threads accessing the same Status must use
+// external synchronization.
+
+// Adapted from Apache Kudu, TensorFlow, Apache Arrow
```
Comment (wgtmac) [inline api/iceberg/status.h]: IIRC, status is the recommended convention from Google coding style. We haven't discussed whether to use status or exception (or something else) across the repo. Personally I prefer using exception which provides better debug-ability and requires less lines to deal with error handling.

File: src/common/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set(ICEBERG_COMMON_SOURCES util/string_builder.cc status.cc)
+set(ICEBERG_COMMON_INCLUDES "${ICEBERG_API_DIR}")
+
+ADD_ICEBERG_LIB(
+  iceberg_common
```
Comment (wgtmac) [inline src/common/CMakeLists.txt]: Why do we need a `iceberg_common` library? Isn't `iceberg-core` the common stuff?

File: CMakeLists.txt (hunk 1 above)
Comment (pitrou) [reply CMakeLists.txt]: Perhaps open a GH issue to discuss it?

File: api/iceberg/util/string_builder.h (hunk 2 above)
Comment (pitrou) [reply api/iceberg/util/string_builder.h]: Either is fine to me.

File: api/iceberg/status.h (hunk 3 above)
Comment (pitrou) [reply api/iceberg/status.h]: Would exceptions allow a rich exception hierarchy together with something similar to `StatusDetail`?

File: src/common/CMakeLists.txt (hunk 4 above)
Comment (zhjwpku) [reply src/common/CMakeLists.txt]: > Why do we need a `iceberg_common` library? Isn't `iceberg-core` the common stuff?

I was thinking that maybe puffin also need the abilities that this module provide? I'm fine if you think this should be moved to iceberg-core.

File: api/iceberg/status.h (hunk 3 above)
Comment (zhjwpku) [reply api/iceberg/status.h]: > IIRC, status is the recommended convention from Google coding style. We haven't discussed whether to use status or exception (or something else) across the repo. Personally I prefer using exception which provides better debug-ability and requires less lines to deal with error handling.

From users' perspective, I think Status is better, developers tend to handle the Status/Result more likely than the exception. The iceberg-rust repo also adopt this kind of design IIUC.

Comment (pitrou) [reply api/iceberg/status.h]: > The iceberg-rust repo also adopt this kind of design IIUC.

Yes, but that's the standard idiom in Rust, and there are syntactical facilities around it that don't exist in C++.
https://doc.rust-lang.org/std/result/

**Comment by gaborkaszab** [comment]
I think this PR makes too much assumptions that we will go with everything that the Arrow community chose. For instance this seems to decide the "exceptions vs status codes" question. Also decides which test framework to use, and additionally pulls in a number of macros too.
I think at this point we should start conversations on these topics, maybe on the Slack channel or as github issues.

I myself planned to come up with the headers for basic classes like Catalog Table and such, but I'd personally prefer exceptions over status codes for many reasons. Method chaining is something I find really powerful in the Java implementation but with status codes you wouldn't be able to do that if you have to check the status for every function call. Not to mention that constantly checking for statuses would make the code less readable. Also Statuses would make the parameter list longer, not to mention that lines would be longer (which somewhat matters now that we went for 90 chars).

Long story short I'm -1 for this PR.

**Comment by zhjwpku** [comment]
> I think this PR makes too much assumptions that we will go with everything that the Arrow community chose. For instance this seems to decide the "exceptions vs status codes" question. Also decides which test framework to use, and additionally pulls in a number of macros too. I think at this point we should start conversations on these topics, maybe on the Slack channel or as github issues.
> 
> I myself planned to come up with the headers for basic classes like Catalog Table and such, but I'd personally prefer exceptions over status codes for many reasons. Method chaining is something I find really powerful in the Java implementation but with status codes you wouldn't be able to do that if you have to check the status for every function call. Not to mention that constantly checking for statuses would make the code less readable. Also Statuses would make the parameter list longer, not to mention that lines would be longer (which somewhat matters now that we went for 90 chars).
> 
> Long story short I'm -1 for this PR.

This PR is not trying to decide which test framework to use, I have started a issue #12 where we can discuss the test framework.

**Comment by zhjwpku** [comment]
> I think this PR makes too much assumptions that we will go with everything that the Arrow community chose. For instance this seems to decide the "exceptions vs status codes" question. Also decides which test framework to use, and additionally pulls in a number of macros too. I think at this point we should start conversations on these topics, maybe on the Slack channel or as github issues.
> 
> I myself planned to come up with the headers for basic classes like Catalog Table and such, but I'd personally prefer exceptions over status codes for many reasons. Method chaining is something I find really powerful in the Java implementation but with status codes you wouldn't be able to do that if you have to check the status for every function call. Not to mention that constantly checking for statuses would make the code less readable. Also Statuses would make the parameter list longer, not to mention that lines would be longer (which somewhat matters now that we went for 90 chars).
> 
> Long story short I'm -1 for this PR.

I have seen some use cases of Method chaining in C++ world, mostly people use it for constructing object I guess. Can you elaborate how we can use it with exception to make the apis easy to use?

I don't like method chaining, it makes debug more difficult, when you set a breakpoint on the method chain and exception happens, how do you know which call give out the exception?

I think it's generally good practice to always write very short and concise lines. Every line should just make one method call.

There is an old stackoverflow question which I think we can take a look, [Method chaining (fluent interfacing) - why is it a good practice, or not?](https://stackoverflow.com/questions/1103985/method-chaining-fluent-interfacing-why-is-it-a-good-practice-or-not)

**Comment by gaborkaszab** [comment]
> There is an old stackoverflow question which I think we can take a look
There is a stackoverflow for everything and for the opposite too. I just googled for "Status codes vs exceptions" and it gives me the first 5 results supporting exceptions over status codes, I guess because the algorithm is biased by my personal preference it learned from me.

> when you set a breakpoint on the method chain and exception happens, how do you know which call give out the exception?
This works well in Java, I don't think it would be any different in C++ world. If you get an exception I guess you can check the trace where it was thrown and then set your debugger accordingly.
Also method chaining doesn't mean that you write everything in a single line, you could break as you want.

I did a quick run through on this PR and my general opinion is this: This PR assumes that we simply want to pull in auxiliary stuff from the Arrow project and without questioning whether we'd need them here or not we just pull entire files into this project. This is pretty difficult to review TBH, and if someone is not from the Arrow project it's even more difficult because they have to figure out the initial motivation of each functionality of another project.

I think this should be the other way around. Let's identify some functionality that we'd like to develop in this project, like interface for catalogs let's say, and if we find something in other projects that could help then pull in only the required part. Or let's have a very simple implementation ourselves, we can still evolve it later on.

**Comment by wgtmac** [comment]
I agree with @gaborkaszab that it would be better to discuss a concrete API design (e.g. Table, FileIO, etc.) before introducing a full-functional status implementation. If we decide to go with status over exception, we can add a basic status class at first and make it powerful later on.

**Comment by zhjwpku** [comment]
> I agree with @gaborkaszab that it would be better to discuss a concrete API design (e.g. Table, FileIO, etc.) before introducing a full-functional status implementation. If we decide to go with status over exception, we can add a basic status class at first and make it powerful later on.

Agreed, maybe we should schedule a sync meeting to discuss some of the issues.

**Comment by lidavidm** [comment]
I'm a bit late here but even if we want to use Status/Result, it may be better to go with a backport of std::expected to at least try to align with the STL.

**Comment by wgtmac** [comment]
@lidavidm I was about to reply to this thread to revive the discussion :)

**Comment by zhjwpku** [comment]
> I'm a bit late here but even if we want to use Status/Result, it may be better to go with a backport of std::expected to at least try to align with the STL.
> 
> Ah sorry, I see discussion moved to #14.

No worry, I totally agree your suggestion, I will work on this way in the following days.

**Comment by lidavidm** [comment]
I have used this before: https://github.com/TartanLlama/expected

I think my main complaint (this applies to Arrow too) is that the structure is a bit opaque in a debugger. Arrow ended up writing a GDB pretty printer script.

File: api/iceberg/util/string_builder.h (hunk 2 above)
Comment (zhjwpku) [reply api/iceberg/util/string_builder.h]: I have changed this to builtin format, let's see if the CI complaints.

File: api/iceberg/status.h (hunk 3 above)
Comment (zhjwpku) [reply api/iceberg/status.h]: We had some discussion that std::expected might be a better solution[1], we haven't decide whether we should adopt C++23 or write a backport version of std::expected. But no matter which way, we need this Status data structure.

[1] https://github.com/apache/iceberg-cpp/issues/14#issuecomment-2581572706

File: src/common/CMakeLists.txt (hunk 4 above)
Comment (zhjwpku) [reply src/common/CMakeLists.txt]: not a problem any more.

File: src/iceberg/status.h [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+  /// Return true iff the status indicates success.
+  constexpr bool ok() const { return (state_ == nullptr); }
+
+  /// Return true iff the status indicates an IO-related failure.
+  constexpr bool IsIOError() const { return code() == StatusCode::IOError; }
+  /// Return true iff the status indicates an unimplemented operation.
+  constexpr bool IsNotImplemented() const { return code() == StatusCode::NotImplemented; }
+  /// Return true iff the status indicates an unknown error.
+  constexpr bool IsUnknownError() const { return code() == StatusCode::UnknownError; }
```
Comment (raulcd) [inline src/iceberg/status.h]:
```suggestion
  /// Return true if the status indicates success.
  constexpr bool ok() const { return (state_ == nullptr); }

  /// Return true if the status indicates an IO-related failure.
  constexpr bool IsIOError() const { return code() == StatusCode::IOError; }
  /// Return true if the status indicates an unimplemented operation.
  constexpr bool IsNotImplemented() const { return code() == StatusCode::NotImplemented; }
  /// Return true if the status indicates an unknown error.
  constexpr bool IsUnknownError() const { return code() == StatusCode::UnknownError; }
```

**Comment by zhjwpku** [comment]
This is obsoleted by #40, though we haven't decided to adopt expected or not, this PR can be closed. Thanks for all the inputs.
